PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,438 @@
+    if (!missingInputEcal_) {
+      for (const auto& photon : thePhotons) {
+        double pt_gam = photon.pt();
+        double eta_gam = photon.eta();
+        if (pt_gam < 20.)
+          continue;
+        if (fabs(eta_gam) < 2.)
+          continue;
+        if (fabs(eta_gam) > 3.)
+          continue;
+        double prefiringprob_gam = getPrefiringRateEcal(eta_gam, pt_gam, h_prefmap_photon_, fluct);
+        nonPrefiringProba[fluct] *= (1. - prefiringprob_gam);
+        nonPrefiringProbaPhoton[fluct] *= (1. - prefiringprob_gam);
+      }
+
+      //Now applying the prefiring maps to jets in the affected regions.
+      for (const auto& jet : theJets) {
+        double pt_jet = jet.pt();
+        double eta_jet = jet.eta();
+        double phi_jet = jet.phi();
+        if (pt_jet < 20.)
+          continue;
+        if (fabs(eta_jet) < 2.)
+          continue;
+        if (fabs(eta_jet) > 3.)
+          continue;
+        if (jetMaxMuonFraction_ > 0 && jet.muonEnergyFraction() > jetMaxMuonFraction_)
+          continue;
+        //Loop over photons to remove overlap
+        double nonprefiringprobfromoverlappingphotons = 1.;
+        bool foundOverlappingPhotons = false;
+        for (const auto& photon : thePhotons) {
+          double pt_gam = photon.pt();
+          double eta_gam = photon.eta();
+          double phi_gam = photon.phi();
+          if (pt_gam < 20.)
+            continue;
+          if (fabs(eta_gam) < 2.)
+            continue;
+          if (fabs(eta_gam) > 3.)
+            continue;
+          double dR2 = reco::deltaR2(eta_jet, phi_jet, eta_gam, phi_gam);
+          if (dR2 > 0.16)
+            continue;
+          double prefiringprob_gam = getPrefiringRateEcal(eta_gam, pt_gam, h_prefmap_photon_, fluct);
+          nonprefiringprobfromoverlappingphotons *= (1. - prefiringprob_gam);
+          foundOverlappingPhotons = true;
+        }
+        //useEMpt =true if one wants to use maps parametrized vs Jet EM pt instead of pt.
+        if (useEMpt_)
+          pt_jet *= (jet.neutralEmEnergyFraction() + jet.chargedEmEnergyFraction());
+        double nonprefiringprobfromoverlappingjet = 1. - getPrefiringRateEcal(eta_jet, pt_jet, h_prefmap_jet_, fluct);
+
+        if (!foundOverlappingPhotons) {
+          nonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet;
+          nonPrefiringProbaJet[fluct] *= nonprefiringprobfromoverlappingjet;
+        }
+        //If overlapping photons have a non prefiring rate larger than the jet, then replace these weights by the jet one
+        else if (nonprefiringprobfromoverlappingphotons > nonprefiringprobfromoverlappingjet) {
+          if (nonprefiringprobfromoverlappingphotons > 0.) {
+            nonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet / nonprefiringprobfromoverlappingphotons;
+            nonPrefiringProbaJet[fluct] *= nonprefiringprobfromoverlappingjet / nonprefiringprobfromoverlappingphotons;
+          } else {
```

Comment:
I'll let @lathomas comment on this one, I'm really not that familiar with the jet part.

Reply:
Hello. I don't have a strong opinion but if a numerical fluctuation brings the non prefiring probability to a negative number, it made sense to me to set it globally to zero.

Reply:
> Hello. I don't have a strong opinion but if a numerical fluctuation brings the non prefiring probability to a negative number, it made sense to me to set it globally to zero.

It makes sense. Also the "photon" one, then. in that case (since you are checking a downward fluctuation of the photon non-prefiring prob).

Reply:
Might it be easier to protect against fluctuations below 0 by changing the `getPrefiringRateMuon` and `getPrefiringRateEcal` methods to not return a rate > 1?

Reply:
Both options are fine with me. In fact this protection already exists for the up/down fluctuation of the ECAL prefiring rate:
https://github.com/cms-sw/cmssw/blob/b87e26fe3f05c6908bfffc8078728d4e4290f0c4/PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc#L376-L380 
Not for the central value though. Maybe even good to have a LogWarning if this happens for the central value?

Reply:
As you can see from that code, I have took the liberty this protection for both the Muon and Ecal cases. I can add LogWarnings for it

Reply:
Done.